Add arXiv data fetching functionality

Goziee-git opened the pull request:
## Fixes
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, please open one before creating this pull request. -->
<!-- Replace [issue number] with the issue number (don't leave the square brackets)--likewise for [issue author]. -->
- Fixes #178 by @Goziee-git

## Description
Implements comprehensive arXiv data collection system to quantify open access academic papers in the commons.

## Type of Change
- [x] New feature implementing data collection from the arXivopen access academic papers
- [x] Data source addition/modification

## Changes Made
- Added arXiv API integration for fetching academic paper metadata using the arXiv API as project requirement for automation of fetching new data sources.
- Implemented data processing pipeline for arXiv submissions in the `scripts/1-fetch/arxiv_fetch.py`
- Created filtering logic for open access and CC-licensed papers 
- Added arXiv data to quarterly reporting system

## Testing
- [x] Static analysis passes (`./dev/check.sh`)
- [x] arXiv API integration tested with sample queries
- [x] Data processing validated with test dataset

## Data Impact
- [x] New data source added (arXiv academic papers)
- [x] Report generation affected (new academic commons metrics)

## Related Documentation
- Updated `sources.md` with arXiv API credentials setup
- Added arXiv processing documentation

## Checklist
<!-- DON'T remove this section or any of the lines. -->
<!-- Leave incomplete or inapplicable lines unchecked. -->
<!-- Replace the [ ] with [x] to check the boxes (there is no space between x and square brackets). -->
- [x] I have read and understood the Developer Certificate of Origin (DCO), below, which covers the contents of this pull request (PR).
- [x] My pull request doesn't include code or content generated with AI.
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the *default* branch of the repository (`main` or `master`).
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin
<!-- YOU MUST READ AND UNDERSTAND THE FOLLOWING ATTESTATION. -->
<!-- -->
<!-- Be aware that copying and pasting from discussion sites or generative AI isn't allowed under this DCO. -->

For the purposes of this DCO, "license" is equivalent to "license or public domain dedication," and "open source license" is equivalent to "open content license or public domain dedication."
<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

Goziee-git commented:
> This is a great start.
> 
> I recommend also developing a data/report plan. For example:
> 
> * It is not meaningful to get a count of a single language (though it is worth noting that other languages are not available).
> * Category codes converted to reporting (words and/or abbreviations instead of acronyms)

@TimidRobot I have removed the query for languages as it returns only English. Also worthy of note here, the arxiv data source accepts papers in other languages but requires that the paper abstracts be submitted in English. so Impossible to get a good distribution of licenses as per language usage.

Also, as suggested. I converted the category codes to reporting words that a more user-friendly and readable using an external `arxiv_category_map.yml`, `in data/2025Q4/1-fetch`. I believe this should make updates repoducible and maintainable over time

Goziee-git commented on arxiv_fetch.py:
> Please remove this file.

@TimidRobot this is the arxiv_fetch.py script youv'e asked me to remove completely😭

Goziee-git replied:
> You have two files named that:
> 
> 1. `arvix_fetch.py`
> 2. `scripts/1-fetch/arvix_fetch.py`
> 
> The 2nd is in the proper location.

@TimidRobot I have seen the error and have removed the file from this PR

Goziee-git commented on scripts/1-fetch/arxiv_fetch.py:
```diff
@@ -0,0 +1,598 @@
+        status_forcelist=[408, 429, 500, 502, 503, 504]
+    )
+    session = requests.Session()
+    session.mount("http://", HTTPAdapter(max_retries=retry_strategy))
```
@TimidRobot I have fixed changes to preserve API calls through HTTPS only

Goziee-git commented on scripts/1-fetch/arxiv_fetch.py:
```diff
@@ -0,0 +1,598 @@
+    )
+    session = requests.Session()
+    session.mount("http://", HTTPAdapter(max_retries=retry_strategy))
+    session.mount("https://", HTTPAdapter(max_retries=retry_strategy))
```
> Please add User-Agent header using `shared.USER_AGENT` (you'll have to make sure your repo/branch is up-to-date to get latest shared library).

sync to upstream from forked repo and also local, rebased changes and add `shared.USER_AGENT` so that the API calls from the script can be properly identified.

Goziee-git commented:
> > Also, as suggested. I converted the category codes to reporting words that a more user-friendly and readable using an external `arxiv_category_map.yml`, `in data/2025Q4/1-fetch`. I believe this should make updates repoducible and maintainable over time.
> 
> * How was `arxiv_category_map.yml` created?
>   
>   * If by script, it should probably go in `dev/`
> * Data that persists should go in `data/` not a specific quarter directory
> 
> > The script now produces `arxiv_2_count_by_category_report.csv` and `arxiv_2_count_by_category_report_agg.csv` for better reporting. Also instead of dumping author count data previously, i implemented a Bucketing approach. in `arxiv_4_count_by_author_bucket.csv` to group author counts into meaningful ranges (1, 2-3, 4-6, 7-10, 11+).
> 
> I'll look at data after outstanding comments are resolved.
> 
> > The script also generates a `arxiv_provenance.json` to record metadata for audit, reproducibility, and provenance
> 
> I'll look at data after outstanding comments are resolved. That said, I'm not excited about adding JSON to the project.

@TimidRobot please as regards the provenance file, i have fixed this issue by changing the output of the scipt from json to yaml so it stays consistent with the project.

Goziee-git commented:
> @Goziee-git Please resolve outstanding conversations.
> 
> Also, you should always review pull requests (PRs) after submitting them. The **Files changed** tab of this PR is full of obvious issues. Also keep in mind that GitHub is a [Foundational technology](https://opensource.creativecommons.org/contributing-code/foundational-tech/) of CC Open Source projects.

@TimidRobot I have deleted the data files showing up from the old commit history and fetch script that was out of the order of the project structure.

Goziee-git commented:
> > Hello @TimidRobot, i observed from multiple results fetched previously that the script failed to fetch CC licenses that may be recorded as hyphenated variants like (CC-BY, CC-BY-NC, etc). I have implemented a compiled regex pattern that replaces the string matching for more robust license detection
> > I have also looked at some of the implementations in other PR to use the normalize_license_text() function for consistent license identification.
> > Please i'ld like to know what your thoughts are on these changes and work continuously on further improvements, Thanks.
> 
> It's probably a good idea to create a function in the shared library eventually. Please leave that to last, however.

@TimidRobot i'ld like to have some clarity on this review. Please do you request that i write a seperate function in the shared.py to normalise licenses, and have the fetch script import the function

Goziee-git commented on data/2025Q4/1-fetch/arxiv_4_count_by_author_count.csv:
> The `PLAN_INDEX` in the `arxiv_4_count_by_author_count.csv` is from the old commit history, i believe the new commit has a resolved version of this issue. Also, when I run the proper script in `scripts/1-fetch/arxiv_fetch.py/arxiv_fetch.py`, it doesn't have the `PLAN_INDEX`. I have also discovered that anytime i make a new commit and push to the forked repo, the data/ files from the initial commit which i have deleted previously as requested are also pushed unintentionally. I am currently investigating this behaviour to see the reason why. Currently, i have created a .gitignore locally to help prevent any unintended push for files that are not staged and committed. While i work on resolving this issue, I will gladly love if you could share a strategy to get over this hurdle

Goziee-git commented:
> > Without feedparser, the script can't process ArXiv's API responses. So i think you have to add the `feedparser` dep to your pipfile to run the script.
> 
> @Goziee-git no, you need to include an updated `Pipfile` and `Pipfile.lock` in your pull request.

@TimidRobot, since this will require that i add a new dependency/tooling to the `Pipefile` and the `Pipfile.lock` would you prefer that i submit the updates for this here without raising an issue for it or otherwise

Babi-B commented on scripts/1-fetch/arxiv_fetch.py:
```diff
@@ -0,0 +1,643 @@
+    PATHS["data_1-fetch"], "arxiv_4_count_by_author_count.csv"
+)
+FILE_ARXIV_AUTHOR_BUCKET = shared.path_join(
+    PATHS["data_1-fetch"], "arxiv_4_count_by_author_bucket.csv"
```
Hi @Goziee-git!

I think you should order your constants

Goziee-git replied:
> Hi @Goziee-git!
> 
> I think you should order your constants

Really, its a good practice to order constants, and i do have alot of constants. I agree with you that ordering them keeps the flow organised and conforms to proper coding practices. This script is still in development and will need all the insights i can get to make it best for longterm use. I appreciate this here @Babi-B, Thanks

Babi-B commented on dev/arxiv_category_converter.py:
```diff
@@ -0,0 +1,73 @@
+        with open(mapping_file, "r") as f:
+            return yaml.safe_load(f) or {}
+    except Exception:
+        return {}
```
Hi @Goziee-git! 

I have a question about this. Why do you choose to silently swallow up the exceptions? Would it not be better to log an error/warning so the user knows the mapping wasn't loaded?

Goziee-git replied:
This script is still in development and thanks @Babi-B for the notes here. I have looked into it carefully and have implemented changes to properly log exceptions instead of silently handling them.

Babi-B commented on dev/arxiv_category_converter.py:
```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env python
```
Is this script meant to be **imported purely as a module**? If it is, I think it is not needed here

Babi-B replied:
I mean the shebang

Babi-B commented on dev/arxiv_category_converter.py:
```diff
@@ -0,0 +1,73 @@
+        data_dir: Directory containing arxiv_category_map.yaml
+    """
+    if not os.path.exists(input_file):
+        return
```
Here too...a silent return. Why not a warning or an exception? Could this not lead to confusing downstream errors?

Babi-B commented on scripts/1-fetch/arxiv_fetch.py:
```diff
@@ -0,0 +1,643 @@
+]
+
+# Log the start of the script execution
+LOGGER.info("Script execution started.")
```
You should put this in the main function so it triggers when the script actually runs

Babi-B commented on scripts/1-fetch/arxiv_fetch.py:
```diff
@@ -0,0 +1,643 @@
+def initialize_data_file(file_path, headers):
+    """Initialize CSV file with headers if it doesn't exist."""
+    if not os.path.isfile(file_path):
+        with open(file_path, "w", newline="") as file_obj:
```
I think you should add `encoding="utf-8"` here too

Babi-B commented on scripts/1-fetch/arxiv_fetch.py:
```diff
@@ -0,0 +1,643 @@
+    retry_strategy = Retry(
+        total=5,
+        backoff_factor=1,
+        status_forcelist=[408, 429, 500, 502, 503, 504],
```
There is a `STATUS_FORCELIST` constant in the `shared.py`

Babi-B commented on scripts/1-fetch/arxiv_fetch.py:
```diff
@@ -0,0 +1,643 @@
+    # author_counts: {license: {author_count(int|None): count}}
+
+    # Save license counts
+    with open(FILE_ARXIV_COUNT, "w", newline="") as fh:
```
`encoding="utf-8"` here too

Babi-B commented on scripts/1-fetch/arxiv_fetch.py:
```diff
@@ -0,0 +1,643 @@
+            writer.writerow({"TOOL_IDENTIFIER": lic, "COUNT": c})
+
+    # Save detailed category counts (code)
+    with open(FILE_ARXIV_CATEGORY, "w", newline="") as fh:
```
`encoding="utf-8"`

Babi-B commented on scripts/1-fetch/arxiv_fetch.py:
```diff
@@ -0,0 +1,643 @@
+        LOGGER.warning("Error loading external arXiv category map: %s", e)
+    results_per_iteration = 50
+
+    search_queries = [
```
All constants are to be moved to the top. They are to be global

Babi-B commented:
Hi @TimidRobot @Goziee-git 

I have noticed entries like these in the CSV files:

```
"TOOL_IDENTIFIER","AUTHOR_COUNT","COUNT"
"UNKNOWN CC legal tool","3","12"
"UNKNOWN CC legal tool","2","18"
"UNKNOWN CC legal tool","1","9"
"UNKNOWN CC legal tool","8","4"
"UNKNOWN CC legal tool","4","13"
"UNKNOWN CC legal tool","10","5"
"UNKNOWN CC legal tool","5","9"
"UNKNOWN CC legal tool","9","4"
```

I was wondering the purpose of collecting ` UKNOWN CC Legal tools` values. Will they be used to diagnose data quality (like a plot of unkown vs known license coverage), or should they be excluded from the analysis phase?

Goziee-git commented on scripts/1-fetch/arxiv_fetch.py:
```diff
@@ -0,0 +1,643 @@
+    return session
+
+
+def normalize_license_text(raw_text: str) -> str:
```
> Please remove the type hint. It can be added back when/if type hints are supported by the project.

@TimidRobot alright this has been done, and implemented, pending approval for the type checker

Goziee-git commented:
> > Hello @TimidRobot, i observed from multiple results fetched previously that the script failed to fetch CC licenses that may be recorded as hyphenated variants like (CC-BY, CC-BY-NC, etc). I have implemented a compiled regex pattern that replaces the string matching for more robust license detection
> > I have also looked at some of the implementations in other PR to use the normalize_license_text() function for consistent license identification.
> > Please i'ld like to know what your thoughts are on these changes and work continuously on further improvements, Thanks.
> 
> It's probably a good idea to create a function in the shared library eventually. Please leave that to last, however.

Hello @TimidRobot, I've been able to apply the logic you recommended here, using a shared library for  a centralized ArXiv category code-to-name conversion system, this way we dont need to use the 

> > Hello @TimidRobot, i observed from multiple results fetched previously that the script failed to fetch CC licenses that may be recorded as hyphenated variants like (CC-BY, CC-BY-NC, etc). I have implemented a compiled regex pattern that replaces the string matching for more robust license detection
> > I have also looked at some of the implementations in other PR to use the normalize_license_text() function for consistent license identification.
> > Please i'ld like to know what your thoughts are on these changes and work continuously on further improvements, Thanks.
> 
> It's probably a good idea to create a function in the shared library eventually. Please leave that to last, however.

Hello @TimidRobot, I've been able to refactor the former logic which used the `arxiv_category_converter.py`script for post processing conversion of the category code to friendly and readable names to a more convenient approach. I have been able to implement the conversion upstream in the `arxiv_fetch.py` script which uses the `shared.normalize_arxiv_category()` when generating data. This way, the old converter script is redundant since we wouldnt be needing it anymore. Please I'ld like to know if you find this approach appropriate and review for changes. Thanks

Goziee-git commented on scripts/1-fetch/arxiv_fetch.py:
```diff
@@ -0,0 +1,735 @@
+# HTTP Retry Configuration
+RETRY_TOTAL = 5
+RETRY_BACKOFF_FACTOR = 1
+
```
@TimidRobot Hello, I have changed the constants and replaced them with literal values in the `arxiv_fetch.py` script as recommended

Goziee-git commented on scripts/1-fetch/arxiv_fetch.py:
```diff
@@ -0,0 +1,735 @@
+        "--limit",
+        type=int,
+        default=DEFAULT_FETCH_LIMIT,
+        help=f"Limit papers to fetch (default: {DEFAULT_FETCH_LIMIT})",
```
@TimidRobot, i have added the requred documentation in the code to reflect the expectations of your review here, also noting that the LIMIT per query set to 500 is an implementation choice to balance against runaway queries and have good data collection

Goziee-git commented:
@TimidRobot Hello, are there any more changes you'll like to be made to the ones already identified and done. Also i'ld like to ask, if there are no more changes to be made on this, what steps would you recommend as the next one for me in the project, with respect to integrating ArXiv as a valid source for the project?

Goziee-git commented on scripts/1-fetch/arxiv_fetch.py:
```diff
@@ -0,0 +1,678 @@
+    # author_counts: {license: {author_count(int|None): count}}
+
+    # Save license counts
+    with open(FILE_ARXIV_COUNT, "w", newline="", encoding="utf-8") as fh:
```
@TimidRobot going over your review on  #183 and #218 i can see you recommend `newline="n\"` across platform, i have implemented this here also to ensure this file remains consistent with the project requirements

Goziee-git commented on scripts/1-fetch/arxiv_fetch.py:
```diff
@@ -0,0 +1,735 @@
+        return "4-6"
+    if 7 <= n <= 10:
+        return "7-10"
+    return "11+"
```
> Where do these buckets come from? Do you have any examples of similar author bucketing?

@TimidRobot, the implementation of the bucketing for the `AUTHOR_COUNT` was because you suggested that it was not so insightful to just merely dump the data, and so i thought a sort of grouping of this data would help us to generate meaningful insights. Raw author counts would create extremely sparse data with many single-occurrence values while Bucketing ("1", "2-3", "4-6", "7-10", "11+") on the other hand, creates meaningful statistical groups for analysis. It also facilitates visualization and reporting by creating compact, aggregated datasets that are easier to process and analyze hence supporting the project's goal of understanding "how knowledge and culture of the commons is distributed".

although these are my thoughts, I am still very happy to know what you think about this. To further share my opinion, these are some helpful links that guided my decision for bucketing the `AUTHOR_COUNT` data.

**Scientometrics Journal Guidelines**: (https://link.springer.com/journal/11192)
• Standard practice in bibliometric studies to group author counts into ranges for collaboration analysis which reduces noise and enables meaningful statistical comparisons across disciplines

**FAIR Data Principles**: (https://www.go-fair.org/fair-principles/)
• Bucketing supports Findability and Reusability by creating standardized categorical data
**NIH Data Management Guidelines**: (https://sharing.nih.gov/data-management-and-sharing-policy)

Goziee-git replied:
I'll like to know if you want me to drop this implementation here, and just dump the `AUTHOR_COUNT` data

Goziee-git commented:
Hello @TimidRobot as per https://github.com/creativecommons/quantifying/pull/179#discussion_r2467105900
the implementation of the bucketing for the AUTHOR_COUNT was because you suggested that it was not so insightful to just merely dump the data, and so i thought a sort of grouping of this data would help us to generate meaningful insights. Raw author counts would create extremely sparse data with many single-occurrence values while Bucketing ("1", "2-3", "4-6", "7-10", "11+") on the other hand, creates meaningful statistical groups for analysis. It also facilitates visualization and reporting by creating compact, aggregated datasets that are easier to process and analyze hence supporting the project's goal of understanding "how knowledge and culture of the commons is distributed".

although these are my thoughts, I am still very happy to know what you think about this. To further share my opinion, these are some helpful links that guided my decision for bucketing the AUTHOR_COUNT data.

Scientometrics Journal Guidelines: (https://link.springer.com/journal/11192)
• Standard practice in bibliometric studies to group author counts into ranges for collaboration analysis which reduces noise and enables meaningful statistical comparisons across disciplines

FAIR Data Principles: (https://www.go-fair.org/fair-principles/)
• Bucketing supports Findability and Reusability by creating standardized categorical data
NIH Data Management Guidelines: (https://sharing.nih.gov/data-management-and-sharing-policy)

Goziee-git commented:
> @Goziee-git please see two unresolved conversations:
> 
> 1. [Add arXiv data fetching functionality #179 (comment)](https://github.com/creativecommons/quantifying/pull/179#discussion_r2467105900)
> 2. [Add arXiv data fetching functionality #179 (comment)](https://github.com/creativecommons/quantifying/pull/179#discussion_r2467123649)

@TimidRobot the conversations mentioned here are identical, please i'ld like to know what your preference is with the `AUTHOR_COUNT`, especially if you want it removed or modified in differnt way
